treq/testing.py

```diff
@@ -205,3 +207,212 @@ def __init__(self, resource):
+    """
+    def __repr__(self):
+        return "ANYTHING"
+
+    def __eq__(self, _):
+        return True
+
+    def __ne__(self, _):
+        return False
+
+ANY = _any()
+
+
+class SequenceStringStubs(object):
+    """
+    Takes a sequence of::
+
+        [((method, url, params, headers, data), (code, headers, body)),
+         ...]
+
+    Expects the requests to arrive in sequence order.  If there are no more
+    responses, or the request's paramters do not match the next item's expected
+    request paramters, raises :obj:`AssertionError`.
+
+    :ivar list sequence: The sequence of expected request arguments mapped to
+        stubbed responses
+    :ivar testcase: A :obj:`TestCase` that can be used to report failures in
+        matching requests against expected requests.  Raising exceptions does
+        not work, because :obj:`Resource.render` will just convert that into
+        a 500 response.
+    """
+    def __init__(self, sequence, testcase):
+        self._sequence = sequence
+        self._testcase = testcase
+
+    def consumed(self):
+        """
+        :return: `bool` representing whether the entire sequence has been
+            consumed.  This is useful in tests to assert that the expected
+            requests have all been made.
+        """
+        return len(self._sequence) == 0
+
+    @contextmanager
+    def consume(self):
+        """
+        Usage::
+
+            stub_treq = StubTreq(StringStubbingResource(
+                sequence_stubs.get_response_for))
+            with sequence_stubs.consume():
+                stub_treq.get('http://fakeurl.com')
+                stub_treq.get('http://another-fake-url.com')
+
+        If there are still remaining expected requests to be made in the
+        sequence, fails the provided test case.
+
+        :return: a context manager that can be used to ensure all expected
+            requests have been made.
+        """
+        yield
+        if not self.consumed():
+            self._testcase.fail("\n".join(
+                ["Not all expected requests were made.  Still expecting:"] +
+                ["- {0}(url={1}, params={2}, headers={3}, data={4})".format(
+                    *expected) for expected, _ in self._sequence]))
+
+    def get_response_for(self, method, url, params, headers, data):
+        """
+        :return: the next response in the sequence, provided that the
+            parameters match the next in the sequence.
+        """
+        if len(self._sequence) == 0:
+            self._testcase.addCleanup(
+                self._testcase.fail,
+                "No more requests expected, but request {0!r} made.".format(
+                    (method, url, params, headers, data)))
+            return (500, {}, "StubbingError")
+
+        expected, response = self._sequence[0]
+        e_method, e_url, e_params, e_headers, e_data = expected
+
+        checks = [
+            (e_method is ANY or e_method.lower() == method.lower(), "method"),
```

Comment:
> I just meant I wanted to use ANY as the sentinel value, and not mix in {} and ANY as a sentinel value.

I think it's better to have an API that doesn't even need a sentinel value, period. I'm not saying that this should be the case at all costs. But if Python code does have a sentinel value, it's almost always None, and if it's not, it's not something that needs to be passed in explicitly by the user (i.e. it may be a default argument to a function that the user can ignore). So here's my kind of hierarchy of desirability with respect to function parameters:

1) no sentinel values needed
2) implicit sentinel value of None
3) explicit sentinel value of None
4) implicit non-None sentinel value
5) explicit non-None sentinel value

I think the API will be easy to understand if described in terms of what checks are done. Right now, the SequenceStringStub docstring doesn't mention `ANY` as a valid value (and if we continue doing this `ANY` identity checks, it definitely should). It also doesn't mention how the tuple elements are interpreted. Let me propose a change.

>    :ivar list sequence: A sequence of tuples of (method, url, params, headers_subset, data), where all elements is checked for equality with the actual request parameters, except for `headers_subset`, which is checked to be a subset of the actual headers. The method should be lowercase.

By reading this docstring, it should be pretty apparent that you can't use ANY for `headers_subset` since it doesn't say it's checked for equality. Even if it's not, and you try to use `ANY`, the test will fail immediately on comparing the headers so it should be pretty easy to realize what the problem is.

Right now the docstring doesn't mention that `ANY` is explicitly checked, so it's already confusing. If I know ANY is an object equal to all values, then I would not expect it to work when passed as `headers_subset`, since I know that headers_subset is *not* checked for equality.

Lastly: if I'm using `mock.ANY`, I would expect to be able to pass it in the same situations as this `ANY` object. So if we do stick with an explicit sentinel, I think it'd be better to call it something else, like MATCH_ALL or something. 

> As for why we are round tripping the expected URL, I was hoping to throw away extra things like query args, a double slash, etc. which URLPath normalizes, but maybe it's better to not do that so its more plain that the user provided the wrong expected value.

Yeah, I agree. If it's test code, it should be pretty precise.

> I think that while we should accept {} for headers though, it would also make sense to still accept ANY? Otherwise it would be the one item in the tuple that ANY can't be used for.

I think it's fine that `ANY` can't be used for headers as long as you stop thinking of `ANY` as an explicit "sentinel value" and more as a "thing that is equal to all things", which is what it's currently documented as.

As a final point, I'm also not totally convinced that we should really be using HasHeaders here. Of course, it would simplify the code to an equality check, which makes this API design simpler, but it also means that users would be able to perform negative tests on their code. If I have an application that *sometimes* requests things with the `X-CreditCardNumber`, then I can very easily see myself writing tests that try to ensure that `X-CreditCardNumber` is *not* passed unless it's strictly needed. glyph says that treq reserves the right to start passing whatever headers it wants, but if it starts passing `X-CreditCardNumber` then I'd also very much like my tests to start failing :)

So if we get rid of the required use of `HasHeaders`, then you get the ability to pass `ANY` as the expected headers field, *and* you get the ability to pass in any other equality-overriding predicate you want, or just a simple, complete value of expected headers.

Okay, that's it. Sorry that this is turning into a multi-volume novel of reviews :(